compiler/rustc_ast_lowering/src/item.rs

```diff
@@ -205,8 +205,38 @@ impl<'hir> LoweringContext<'_, 'hir> {
                 sig: FnSig { decl, header, span: fn_sig_span },
                 generics,
                 body,
+                contract,
                 ..
             }) => {
+                if let Some(contract) = contract {
+                    assert!(self.contract.is_none());
+                    let requires = contract.requires.clone();
+                    let ensures = contract.ensures.clone();
+                    let ensures = if let Some(ens) = ensures {
+                        // FIXME: this needs to be a fresh (or illegal) identifier to prevent
+                        // accidental capture of a parameter or global variable.
+                        let checker_ident: Ident =
+                            Ident::from_str_and_span("__ensures_checker", ens.span);
+                        let (checker_pat, checker_hir_id) = self.pat_ident_binding_mode_mut(
+                            ens.span,
+                            checker_ident,
+                            hir::BindingMode::NONE,
+                        );
+
+                        Some(crate::FnContractLoweringEnsures {
+                            expr: ens,
+                            fresh_ident: (checker_ident, checker_pat, checker_hir_id),
+                        })
+                    } else {
+                        None
+                    };
+
+                    self.contract.replace(crate::FnContractLoweringInfo {
+                        span,
+                        requires,
+                        ensures,
+                    });
```

Comment:
The previous value needs to be reinstated after the body is finished lowering, otherwise functions will end up losing the contract info after lowering a nested function that also has contracts. Similarly nested functions without contract info will end up picking up parent contracts

Reply:
Ack, good catch. Let me make some tests of those scenarios and then fix this.

Reply:
So I'm feeling dense:  I agree with you that it seems like this should be necessary. 

But so far my attempts to witness a failure of the form you describe have come up short. (E.g. I tried making test cases where the AST has nested function items with and without contracts, and also closure expressions (which cannot currently have contracts attached in this system, but in theory could run afoul of the problem that you outlined).

To be concrete, here one iteration of the test case I've been trying to develop to show the problem (where the idea is to show that either an outer contract is erroneously applied to an inner item, or that an inner item's contract is erroneously applied to an outer item). But this has not managed to flag an issue against my PR, which has been compiling and running the code below without complaint (and when I deliberately add contract check failures, they fail, so its not a case of missing a check failure).

Do you have any suggestions for what construct I should be using to test this scenario? (At one point I thought that maybe my use of `self.contract.take()` in `lower_fn_body` might explain how I'm sidestepping problems here, but I think that `take()` call happens too late to explain anything at all.)

```rust
//@ run-pass
//@ compile-flags: -Zcontract-checks=yes
#![feature(rustc_contracts)]
#![allow(dead_code, unused_variables)]

#[core::contracts::requires(x.outer > 0)]
fn outer(x: Outer) {
    #[core::contracts::requires(x.inner1 < 0)]
    fn inner1(x: Inner1) {

    }
    fn inner2(x: Inner2) {

    }
    let _inner_closure = |x: Inner2| {
    };
    inner1(Inner1 { inner1: -x.outer });
    inner2(Inner2 { inner2: 0 });
    _inner_closure(Inner2 { inner2: -x.outer });
}

fn main() {
    outer(Outer { outer: 1 });
}

struct Outer { outer: i32 }
struct Inner1 { inner1: i32 }
struct Inner2 { inner2: i32 }
```

Reply:
(I'm nonetheless happy to make the suggested change, because I believe it will be more robust than what i had previously. I'm just not thrilled that I cannot yet demonstrate the *necessity* of the change.)

Reply:
(and *of course* I should also make test cases for this and many similar cases of nesting)

Reply:
Aha! 

I found an example that illustrates the problem!

The contracts are indeed being erroneously inherited by the closure (and so e.g. the pre-condition for the outer function is run again when the closure is invoked).

However, my earlier attempt to show the problem was flawed, in that I had thought that I could illustrate the problem by injecting shadowing bindings. Our resolution system is too smart for that and was keeping the contract checks associated with the *outer* bindings.

The concrete way I illustrated this happening is by mutating the parameter state so that if the contract check gets run again after we have entered the body, it will fail, as shown here:

```rust
//@ compile-flags: -Zcontract-checks=yes
#![feature(rustc_contracts)]

struct Outer { outer: std::cell::Cell<i32> }

#[core::contracts::requires(x.outer.get() > 0)]
fn outer(x: Outer) {
    let inner_closure = || { };
    x.outer.set(0);
    inner_closure();
}

fn main() {
    outer(Outer { outer: 1.into() });
}
```

Reply:
😆 oh right, only nested bodies, not nested items have this issue. I wonder how const blocks would have fared :D

Reply:
Oh I'll add a test for those too, now that I know the kinds of patterns I need to check.

Reply:
(the problem itself is now fixed; I merged the fix into 190f93df9c18de755ae8e405923575b1e0893dfe )

Reply:
Okay I think const blocks were never vulnerable to the problem with precondition checks, because the code path I'm modifying is `lower_fn_body`, but const blocks go straight to `lower_body`. 

As for post-condition checks ... is there any way to make a `return`-expression part of a const body (apart from nesting closures within it)? My experiments indicate that immediately runs into [E0572](https://doc.rust-lang.org/error_codes/E0572.html), so I think post-condition checks were also safe.

Having said that, I'll still try to make some defensive tests. I just cannot make them exercise the same paths that the closure-oriented tests did, so I won't be able to actually demonstrate the bug via const blocks.

Reply:
Yea, const blocks seem to avoid the issues by construction.